File: tests/mozmill/tests/testLinks/testTextSelection/testOpenInNewTab.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

var rpRootDir = "../../../";
var rpConst = require(rpRootDir + "lib/constants");
var rootDir = rpRootDir + rpConst.mozmillTestsRootDir;

var {assert, expect} = require(rootDir + "lib/assertions");
var prefs = require(rootDir + "lib/prefs");
var tabs = require(rootDir + "firefox/lib/tabs");
var utils = require(rootDir + "lib/utils");

var rpUtils = require(rpRootDir + "lib/rp-utils");

const TEST_URL = "http://www.maindomain.test/link_1.html";


var setupModule = function(aModule) {
  aModule.controller = mozmill.getBrowserController();
  aModule.tabBrowser = new tabs.tabBrowser(aModule.controller);
  aModule.tabBrowser.closeAllTabs();

  prefs.setPref(rpConst.PREF_DEFAULT_ALLOW, false);
}

var teardownModule = function(aModule) {
  prefs.clearUserPref(rpConst.PREF_DEFAULT_ALLOW);
  utils.closeContentAreaContextMenu(aModule.controller);
  aModule.tabBrowser.closeAllTabs();
}


var testOpenInNewTab = function() {
  var tabIndex = tabBrowser.selectedIndex;

  controller.open(TEST_URL);
  controller.waitForPageLoad();

  let textURL = findElement.ID(controller.window.document, "text_url_1");
  let textURLValue = textURL.getNode().textContent;

  rpUtils.selectText(tabBrowser, textURLValue);

  // perform right-click and entry selection
  var contextMenu = controller.getMenu("#contentAreaContextMenu");
  contextMenu.select("#context-openlinkintab", textURL);

  assert.waitFor(() => { return tabBrowser.length == 2; }, "New Tab opened.");

  rpUtils.waitForTabLoad(controller, tabBrowser.getTab(tabIndex));

  assertNoRedirects();
  assertCorrectLocations(textURLValue);
}

var assertCorrectLocations = function(linkURL) {
  for (let index = 1; index < tabBrowser.length; ++index) {
    tabBrowser.selectedIndex = index;
    assert.equal(controller.tabs.activeTab.location.href, linkURL,
        "The location in the new tab is correct.");
  }
}

/**
 * Assert that the link clicks have not been detected as redirects.
 */
var assertNoRedirects = function() {
  for (let index = 0; index < tabBrowser.length; ++index) {
    var panel = tabBrowser.getTabPanelElement(index,
        '/{"value":"' + rpConst.REDIRECT_NOTIFICATION_VALUE + '"}');
    assert.ok(false === panel.exists(),
        "Following the link didn't cause a redirect");
  }
}

File: tests/marionette/tests/links/text_selection/test_open_in_new_tab.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from rp_ui_harness import RequestPolicyTestCase


TEST_URL = "http://www.maindomain.test/link_1.html";
PREF_DEFAULT_ALLOW = "extensions.requestpolicy.defaultPolicy.allow"


class TestOpenInNewTab(RequestPolicyTestCase):

    def setUp(self):
        RequestPolicyTestCase.setUp(self)
        self.prefs.set_pref(PREF_DEFAULT_ALLOW, False);


    def test_open_in_new_tab(self):
        with self.marionette.using_context("content"):
            # load the test url
            self.marionette.navigate(TEST_URL)

            # find the URL string and its wrapping element
            url_wrapper = self.marionette.find_element("id", "text_url_1")
            url = url_wrapper.get_attribute("textContent")

            # select the URL
            self.web_utils.select_element_text(url_wrapper)

        with self.marionette.using_context("chrome"):
            # perform right-click and entry selection
            self.ctx_menu.select_entry("context-openlinkintab", url_wrapper)

            self.assertEqual(len(self.browser.tabbar.tabs), 2,
                             "A new tab has been opened.")

            [origin_tab, dest_tab] = self.browser.tabbar.tabs
            self.tabs.wait_until_loaded(dest_tab)

            # checks in the origin's tab
            origin_tab.switch_to()
            self.assertFalse(self.redir.panel_exists(),
                             ("Following the link didn't cause a "
                              "redirect in the origin tab."))

            # checks in the destination's tab
            dest_tab.switch_to()
            self.assertEqual(dest_tab.location, url,
                             "The location in the new tab is correct.")
            self.assertFalse(self.redir.panel_exists(),
                             ("Following the link didn't cause a "
                              "redirect in the destination tab."))

            dest_tab.close()
